proc/combiner.go

```diff
@@ -0,0 +1,127 @@
+package proc
+
+import (
+	"io"
+
+	"github.com/brimsec/zq/expr"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zng"
+)
+
+type MergeFunc func(*zng.Record, ...*zng.Record) (*zng.Record, error)
+
+// Combiner reads from two or more sorted inputs, implementing a new
+// zbuf.Reader from the inputs. It uses a comparison function to
+// merge-sort its outputs. If inputs are all sorted according to the
+// same comparison function, the combiner output will similarly be
+// sorted.
+//
+// If the combiner's MergeFunc is non-nil, that function is invoked to
+// merge all identical records into one upon reading from the
+// combiner, where record identity is defined by the comparison
+// function. If it is nil, then a default merger is used that returns
+// the first record of each identical set. This can be used to
+// deduplicate records.
+type Combiner struct {
+	merge   MergeFunc
+	compare expr.CompareFn
+	readers []zbuf.Reader
+	done    []bool
+	hol     []*zng.Record
+}
+
+// NewCombiner returns a new combiner
+func NewCombiner(readers []zbuf.Reader, c expr.CompareFn, m MergeFunc) *Combiner {
```

Comment:
Instead of a creating another version of Combiner why not use sort's `runManager` and make a separate zbuf.Reader that merges adjacent equal records? Otherwise feels like we're adding a lot of code that already exists elsewhere.

Reply:
(Nit: this isn't another version of Combiner as it replaces the zdx one. Also, as noted above, I think this one can be unified with the single other one, which is currently in `zbuf`).

Good question. One notable difference is that `runManager` always waits for all records to be written before it starts to read. Here, we need to be able to start reading before all records are written, which leads to `combiner.PeekMin()` and `combiner.AddReader()`. I don't think those can be easily added to runManager.

With that said, your point about having merging take place outside of the combiner is a good one. I thought of that along the way but went with the path of least resistance which was to do it the way zdx.Combiner did. But we can change that if it's nicer.

Reply:
> With that said, your point about having merging take place outside of the combiner is a good one.

I've done this in eb1d845d9d0e2978aabbd6d01a2f034bb77854a6.